src/ev-matrix.c

```diff
@@ -596,6 +605,35 @@ ev_matrix_remove_pusher (GStrv args, GError **err)
 }
 
 
+static GString *
+ev_matrix_join_room (GStrv args, GError **err)
+{
+  const char *room;
+
+  g_assert (CM_IS_CLIENT (client));
+
+  if (g_strv_length (args) < 1) {
+    g_set_error (err, G_IO_ERROR, G_IO_ERROR_FAILED, "Not enough arguments");
+    return NULL;
+  }
+
+  if (g_strv_length (args) > 1) {
+    g_set_error (err, G_IO_ERROR, G_IO_ERROR_FAILED, "Too many arguments");
+    return NULL;
+  }
+
+  room = args[0];
+  if (!g_regex_match (room_regex, room, G_REGEX_MATCH_DEFAULT, NULL)) {
+    g_set_error (err, G_IO_ERROR, G_IO_ERROR_FAILED, "Not a valid room id or alias");
+    return NULL;
+  }
+
+  if (!cm_client_join_room_sync (client, room, err))
+    return g_string_new_take (g_strdup_printf ("Failed to join '%s': %s", room, (*err)->message));
```

Comment:
I'd rather set `err` here like e.g.:

```
g_set_error (err, G_IO_ERROR, G_IO_ERROR_FAILED, "No pushers - did you run /get-pushers ?");
```

Reply:
Ah, my thinking here was that `cm_client_join_room_sync()` would already give us a meaningful error
(it's semi meaningful now, i.e. no proper error domain/code)
like "You need an invite", "You have been banned" etc

As a data point this is how it currently looks like

```
/join #music:fortysixandtwo.eu
(eigenvalue:629747): GLib-GIO-DEBUG: 19:47:11.193: g_socket_client_connect_async: starting new g_socket_client_connect_async() with GTask 0x555555d004e0 and GSocketClientAsyncConnectData 0x5555555eaab0
(eigenvalue:629747): GLib-GIO-DEBUG: 19:47:11.193: GSocketClient: Starting new address enumeration
(eigenvalue:629747): pxbackend-DEBUG: 19:47:11.193: px_manager_get_proxies_sync: url=https://matrix.fortysixandtwo.eu:443 online=1
(eigenvalue:629747): pxbackend-DEBUG: 19:47:11.193: px_manager_get_proxies_sync: Proxy[0] = direct://
(eigenvalue:629747): GLib-GIO-DEBUG: 19:47:11.194: GSocketClient: Address enumeration succeeded
(eigenvalue:629747): GLib-GIO-DEBUG: 19:47:11.194: g_socket_client_enumerator_callback: starting connection attempt 0x5555559a6f70 for GSocketClientAsyncConnectData 0x5555555eaab0
(eigenvalue:629747): GLib-GIO-DEBUG: 19:47:11.194: GSocketClient: Starting TCP connection attempt
(eigenvalue:629747): GLib-GIO-DEBUG: 19:47:11.219: GSocketClient: TCP connection successful
(eigenvalue:629747): GLib-GIO-DEBUG: 19:47:11.219: GSocketClient: Starting application layer connection
(eigenvalue:629747): GLib-GIO-DEBUG: 19:47:11.219: GSocketClient: Connection successful!
(eigenvalue:629747): cm-net-DEBUG: 19:47:11.343: Error loading from stream: You are not invited to this room.

Failed to join '#music:fortysixandtwo.eu': You are not invited to this room.
```

Reply:
> Ah, my thinking here was that `cm_client_join_room_sync()` would already give us a meaningful error

You can sure build a nice error message but you should return it as a `GError`.

Reply:
That is what I mean:
We already get a `GError` from `cm_client_join_room_sync()`, and after rechecking the code
it should already be specific/meaningful enough (sorry for the confusion when I claimed earlier that it probably wasn't meaningful).

In the case from above we *should* get a `CM_ERROR_FORBIDDEN` already.

Reply:
I think I explain myself badly: If you want to use the error message from the error in `err` verbatim, then just return `NULL` as string. The upper layers will then use the filled in error.  If you return non-`NULL` then `run_command` will think: `success` (so let's use that string) and ignore what is in `error`. (It's basically an API error to return a `GString` and set `GError`). This is the usual gilb pattern: return `NULL`/`FALSE` and use `error` for the error details (libcmatrix does that often a bit differently and return `NULL` even in the non-error case.

Reply:
ah, that makes sense (I didn't look at how the commands actually work). thanks for ELI5 :)